polynote-frontend/polynote/ui/component/notebook.ts

```diff
@@ -245,21 +251,44 @@ export class NotebookUI extends UIMessageTarget {
                 this.queuedCells = this.queuedCells.filter(item => item !== cellId)
             } else if (status === TaskStatus.Complete || status === TaskStatus.Error) {
                 this.runningCell = undefined;
-                this.queuedCells = this.queuedCells.filter(item => item !== cellId) // just in case
+                this.queuedCells = this.queuedCells.filter(item => item !== cellId); // just in case
+
+                // only notify if this notebook doesn't have focus
+                if (!this.isFocused()) {
+                    Notification.requestPermission().then((result) => {
```

Comment:
What if they don't want notifications? Is this going to ask them for permission every time a cell completes? The way I've seen this done before is that there's one place where you can click a button and get the permission prompt.

Also, if they do a "Run All" or similar, they probably don't want notifications for every cell that finishes; only if there's an error or the whole notebook is finished running.

Reply:
When I tested this a couple weeks ago it looked like it only requested the permission one time. But it's a good point to move that permission request somewhere else (into the settings perhaps). 

Right now the way it works is that only one notification is active at a given time, so you get a notification when a cell finishes and then that notification is updated when another cell finishes. If you close the first notification you will get another one. 

I agree that we probably don't want to spam users when they hit Run All or Run To Cursor. I can make the change to only notify on completion / error.

Reply:
Yep, at least in Chrome it asks you once and saves your setting under Site Settings.